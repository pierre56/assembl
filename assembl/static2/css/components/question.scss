.question-footer {
  .button-light {
    color: $first-color;

    &:hover,
    &:focus,
    &:active,
    &:focus:active,
    &:active:hover,
    &[disabled]:hover {
      color: #fff; /* opposite of current non-hover color */
    }
  }
}

.back-btn-container {
  margin-top: 105px;

  .button-dark {
    color: $first-color;

    &:hover,
    &:focus,
    &:active,
    &:focus:active,
    &:active:hover,
    &[disabled]:hover {
      color: #fff; /* opposite of current non-hover color */
    }
  }
}
