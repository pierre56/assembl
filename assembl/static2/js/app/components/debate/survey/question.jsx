// @flow
import * as React from 'react';
import { graphql, compose } from 'react-apollo';
import { connect } from 'react-redux';
import { Translate, I18n } from 'react-redux-i18n';
import { Grid, Col, Button } from 'react-bootstrap';
import { EditorState } from 'draft-js';

import { getConnectedUserId, getPostPublicationState } from '../../../utils/globalFunctions';
import { inviteUserToLogin, displayAlert } from '../../../utils/utilityManager';
import { connectedUserIsAdmin } from '../../../utils/permissions';
import { SMALL_SCREEN_WIDTH, MINIMUM_BODY_LENGTH } from '../../../constants';
import { withScreenDimensions } from '../../common/screenDimensions';
import RichTextEditor from '../../common/richTextEditor';
import { convertEditorStateToHTML } from '../../../utils/draftjs';
// graphql
import createPostMutation from '../../../graphql/mutations/createPost.graphql';
import DiscussionPreferencesQuery from '../../../graphql/DiscussionPreferencesQuery.graphql';

type Props = {
  isDebateModerated: boolean,
  isPhaseCompleted: boolean,
  title: string,
  contentLocale: string,
  questionId: string,
  scrollToQuestion: Function,
  index: number,
  refetchTheme: Function,
  mutate: Function,
  screenHeight: number,
  screenWidth: number,
  questionsLength: number
};

type State = {
  buttonDisabled: boolean,
  postBody: EditorState
};

export class Question extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      buttonDisabled: false,
      postBody: EditorState.createEmpty()
    };
  }

  createPost = () => {
    const { contentLocale, questionId, scrollToQuestion, index, refetchTheme, isDebateModerated } = this.props;
    const body = this.state.postBody;
    const userIsAdmin = connectedUserIsAdmin();
    const publicationState = getPostPublicationState(isDebateModerated, userIsAdmin);
    this.setState({ buttonDisabled: true }, () =>
      this.props
        .mutate({
          variables: {
            contentLocale: contentLocale,
            ideaId: questionId,
            body: convertEditorStateToHTML(body),
            publicationState: publicationState
          }
        })
        .then(() => {
          scrollToQuestion(true, index + 1);
          const successMessage = isDebateModerated && !userIsAdmin ? 'postToBeValidated' : 'postSuccess';
          displayAlert('success', I18n.t(`debate.survey.${successMessage}`));
          refetchTheme();
          this.setState({
            postBody: EditorState.createEmpty(),
            buttonDisabled: false
          });
        })
        .catch((error) => {
          displayAlert('danger', error.message);
          this.setState({
            buttonDisabled: false
          });
        })
    );
  };

  updateBody = (newValue: EditorState) => {
    this.setState({
      postBody: newValue
    });
  };

  redirectToLogin = () => {
    const isUserConnected = getConnectedUserId();
    const { scrollToQuestion, index } = this.props;
    if (!isUserConnected) {
      inviteUserToLogin();
    } else {
      scrollToQuestion(true, index);
    }
  };

  getPostBodyCharCount = () => {
    const { postBody } = this.state;
    if (postBody) {
      return postBody.getCurrentContent().getPlainText().length;
    }

    return 0;
  };

  render() {
    const { index, isPhaseCompleted, title, screenWidth, screenHeight, questionsLength } = this.props;
    const questionTitle = questionsLength > 1 ? `${index}/ ${title}` : title;
    let height = screenHeight;
    const timelineElm = document && document.getElementById('timeline');
    // This is necessary to bypass an issue with Flow
    if (timelineElm) {
      height = screenHeight - timelineElm.clientHeight;
    }
    return (
      <section
        className={isPhaseCompleted ? 'hidden' : 'questions-section'}
        id={`q${index}`}
        style={screenWidth >= SMALL_SCREEN_WIDTH ? { height: height } : { height: '100%' }}
      >
        <Grid fluid className="background-grey">
          <div className="max-container">
            <div className="question-title">
              <div className="title-hyphen">&nbsp;</div>
              <h1 className="dark-title-5">{questionTitle}</h1>
            </div>
            <Col xs={12} md={9} className="col-centered">
              <RichTextEditor
                editorState={this.state.postBody}
                maxLength={1000}
                onChange={this.updateBody}
                placeHolder={I18n.t('debate.survey.txtAreaPh')}
                handleInputFocus={this.redirectToLogin}
              />
              {this.getPostBodyCharCount() > MINIMUM_BODY_LENGTH && (
                <Button
                  onClick={this.createPost}
                  disabled={this.state.buttonDisabled}
                  className="button-submit button-dark right margin-l clear"
                >
                  <Translate value="debate.survey.submit" />
                </Button>
              )}
            </Col>
          </div>
        </Grid>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  contentLocale: state.i18n.locale
});

export default compose(
  connect(mapStateToProps),
  graphql(createPostMutation),
  graphql(DiscussionPreferencesQuery, {
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      return {
        error: data.error,
        loading: data.loading,
        isDebateModerated: data.discussionPreferences.withModeration
      };
    }
  }),
  withScreenDimensions
)(Question);