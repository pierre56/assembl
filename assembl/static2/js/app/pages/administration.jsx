import React from 'react';
import { Translate } from 'react-redux-i18n';
import { compose, graphql } from 'react-apollo';
import { filter } from 'graphql-anywhere';
import { Grid, Row, Col } from 'react-bootstrap';
import { connect } from 'react-redux';
import moment from 'moment';

import { displayLanguageMenu } from '../actions/adminActions';
import { updateVoteSessionPage, updateVoteModules, updateVoteProposals } from '../actions/adminActions/voteSession';
import { updatePhases } from '../actions/adminActions/timeline';
import { updateSections } from '../actions/adminActions/adminSections';
import { updateLandingPageModules } from '../actions/adminActions/landingPage';
import { updateTextFields } from '../actions/adminActions/profileOptions';
import manageErrorAndLoading from '../components/common/manageErrorAndLoading';
import mergeLoadingAndError from '../components/common/mergeLoadingAndError';
import Menu from '../components/administration/menu';
import LanguageMenu from '../components/administration/languageMenu';
import SectionsQuery from '../graphql/SectionsQuery.graphql';
import TextFields from '../graphql/TextFields.graphql';
import LegalContentsQuery from '../graphql/LegalContents.graphql';
import VoteSessionQuery from '../graphql/VoteSession.graphql';
import LandingPageModules from '../graphql/LandingPageModules.graphql';
import TimelineQuery from '../graphql/Timeline.graphql';
import { convertEntriesToEditorState } from '../utils/draftjs';
import { getPhaseId } from '../utils/timeline';
import { fromGlobalId } from '../utils/globalFunctions';
import { addEnumSuffixToModuleTitles } from '../components/administration/landingPage/addEnumSuffixToModuleTitles';

const SECTIONS_WITHOUT_LANGUAGEMENU = ['1', '6'];

class Administration extends React.Component {
  constructor(props) {
    super(props);
    this.putVoteSessionInStore = this.putVoteSessionInStore.bind(this);
    this.putLandingPageModulesInStore = this.putLandingPageModulesInStore.bind(this);
    this.putTextFieldsInStore = this.putTextFieldsInStore.bind(this);
    this.state = {
      showLanguageMenu: true
    };
  }

  componentDidMount() {
    // we need to use the redux store for administration data to be able to use a
    // "global" save button that will do all the mutations "at once"
    this.putSectionsInStore(this.props.sections);
    this.putVoteSessionInStore(this.props.voteSession);
    this.putVoteModulesInStore(this.props.voteSession);
    this.putVoteProposalsInStore(this.props.voteSession);
    const isHidden =
      this.props.identifier === 'discussion' && SECTIONS_WITHOUT_LANGUAGEMENU.includes(this.props.location.query.section);
    this.props.displayLanguageMenu(isHidden);
    this.putLandingPageModulesInStore(this.props.landingPageModules);
    this.putTextFieldsInStore(this.props.textFields);
    this.putTimelinePhasesInStore(this.props.timeline);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.sections !== this.props.sections) {
      this.putSectionsInStore(nextProps.sections);
    }

    if (nextProps.voteSession !== this.props.voteSession) {
      this.putVoteSessionInStore(nextProps.voteSession);
      this.putVoteModulesInStore(nextProps.voteSession);
      this.putVoteProposalsInStore(nextProps.voteSession);
    }

    if (nextProps.landingPageModules !== this.props.landingPageModules) {
      this.putLandingPageModulesInStore(nextProps.landingPageModules);
    }

    const isHidden =
      nextProps.identifier === 'discussion' && SECTIONS_WITHOUT_LANGUAGEMENU.includes(nextProps.location.query.section);
    this.props.displayLanguageMenu(isHidden);

    if (nextProps.textFields !== this.props.textFields) {
      this.putTextFieldsInStore(nextProps.textFields);
    }

    if (nextProps.timeline !== this.props.timeline) {
      this.putTimelinePhasesInStore(nextProps.timeline);
    }
  }

  putVoteSessionInStore(voteSession) {
    const emptyVoteSession = {
      id: '',
      titleEntries: [],
      seeCurrentVotes: false,
      subTitleEntries: [],
      instructionsSectionTitleEntries: [],
      instructionsSectionContentEntries: [],
      propositionsSectionTitleEntries: [],
      headerImage: {
        externalUrl: '',
        mimeType: '',
        title: ''
      },
      publicVote: true,
      modules: []
    };
    const filteredVoteSession = filter(VoteSessionQuery, { voteSession: voteSession || emptyVoteSession });
    const voteSessionForStore = {
      ...filteredVoteSession.voteSession,
      instructionsSectionContentEntries: filteredVoteSession.voteSession.instructionsSectionContentEntries
        ? convertEntriesToEditorState(filteredVoteSession.voteSession.instructionsSectionContentEntries)
        : null
    };
    this.props.updateVoteSessionPage(voteSessionForStore);
  }

  putTimelinePhasesInStore(timeline) {
    if (timeline) {
      const filteredPhases = filter(TimelineQuery, { timeline: timeline });
      const phasesForStore = filteredPhases.timeline.map(phase => ({
        ...phase,
        start: moment(phase.start),
        end: moment(phase.end)
      }));
      this.props.updatePhases(phasesForStore);
    }
  }

  putVoteModulesInStore(voteSession) {
    const filteredVoteSession = filter(VoteSessionQuery, { voteSession: voteSession });
    const modules =
      filteredVoteSession.voteSession && filteredVoteSession.voteSession.modules ? filteredVoteSession.voteSession.modules : [];
    this.props.updateVoteModules(modules);
  }

  putVoteProposalsInStore(voteSession) {
    const proposals = voteSession ? filter(VoteSessionQuery, { voteSession: voteSession }).voteSession.proposals : [];
    const proposalsForStore = proposals.map(proposal => ({
      ...proposal,
      descriptionEntries: proposal.descriptionEntries ? convertEntriesToEditorState(proposal.descriptionEntries) : null
    }));
    this.props.updateVoteProposals(proposalsForStore);
  }

  putSectionsInStore(sections) {
    if (sections) {
      const filteredSections = filter(SectionsQuery, {
        sections: sections.filter(section => section.sectionType !== 'ADMINISTRATION')
      });
      this.props.updateSections(filteredSections.sections);
    }
  }

  putLandingPageModulesInStore(landingPageModules) {
    if (landingPageModules) {
      const filtered = filter(LandingPageModules, { landingPageModules: landingPageModules });
      const landingPageModulesWithUpdatedTitles = addEnumSuffixToModuleTitles(filtered.landingPageModules);
      this.props.updateLandingPageModules(landingPageModulesWithUpdatedTitles);
    }
  }

  putTextFieldsInStore(textFields) {
    if (textFields) {
      const filtered = filter(TextFields, { textFields: textFields });
      this.props.updateTextFields(filtered.textFields);
    }
  }

  render() {
    const {
      children,
      locale,
      params,
      refetchTabsConditions,
      refetchSections,
      refetchLegalContents,
      refetchVoteSession,
      refetchLandingPageModules,
      refetchTextFields,
      refetchTimeline,
      timeline
    } = this.props;
    const { phase } = params;
    const phaseId = getPhaseId(timeline, phase);
    const discussionPhaseId = phaseId ? fromGlobalId(phaseId) : null;
    const childrenWithProps = React.Children.map(children, child =>
      React.cloneElement(child, {
        discussionPhaseId: discussionPhaseId,
        locale: locale,
        refetchTabsConditions: refetchTabsConditions,
        refetchVoteSession: refetchVoteSession,
        refetchSections: refetchSections,
        refetchLandingPageModules: refetchLandingPageModules,
        refetchLegalContents: refetchLegalContents,
        refetchTextFields: refetchTextFields,
        refetchTimeline: refetchTimeline
      })
    );
    return (
      <div className="administration">
        <div className="save-bar">
          <div className="max-container">
            <Grid fluid>
              <Row>
                <Col xs={12} md={3} />
                <Col xs={12} md={8}>
                  {/* save button is moved here via a ReactDOM portal */}
                  <div id="save-button" />
                </Col>
                <Col xs={12} md={1} />
              </Row>
            </Grid>
          </div>
        </div>
        <div className="max-container">
          <Grid fluid>
            <Row>
              <Col xs={12} md={3}>
                <div className="admin-menu-container">
                  <Menu timeline={timeline} locale={locale} requestedPhase={phase} />
                </div>
              </Col>
              <Col xs={12} md={8}>
                {!timeline ? (
                  <div>
                    <Translate value="administration.noTimeline" />
                  </div>
                ) : null}
                {childrenWithProps}
              </Col>
              <Col xs={12} md={1}>
                <LanguageMenu />
              </Col>
            </Row>
          </Grid>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  timeline: state.timeline,
  locale: state.i18n.locale
});

const mapDispatchToProps = dispatch => ({
  updateSections: sections => dispatch(updateSections(sections)),
  updateVoteModules: voteModules => dispatch(updateVoteModules(voteModules)),
  updateVoteSessionPage: voteSession => dispatch(updateVoteSessionPage(voteSession)),
  updateVoteProposals: voteProposals => dispatch(updateVoteProposals(voteProposals)),
  displayLanguageMenu: isHidden => dispatch(displayLanguageMenu(isHidden)),
  updateLandingPageModules: landingPageModules => dispatch(updateLandingPageModules(landingPageModules)),
  updateTextFields: textFields => dispatch(updateTextFields(textFields)),
  updatePhases: phases => dispatch(updatePhases(phases))
});

const isNotInAdminSection = adminSectionName => props => !props.router.getCurrentLocation().pathname.endsWith(adminSectionName);

const isNotInDiscussionAdmin = isNotInAdminSection('discussion');
const isNotInLandingPageAdmin = isNotInAdminSection('landingPage');

export default compose(
  connect(mapStateToProps, mapDispatchToProps),
  graphql(VoteSessionQuery, {
    skip: ({ params, timeline }) => !timeline || params.phase !== 'voteSession',
    options: ({ locale, timeline, params }) => {
      const phaseId = getPhaseId(timeline, params.phase);
      const discussionPhaseId = phaseId ? fromGlobalId(phaseId) : null;
      return {
        variables: { discussionPhaseId: discussionPhaseId, lang: locale }
      };
    },
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          voteSessionMetadata: {
            error: data.error,
            loading: data.loading
          }
        };
      }

      return {
        voteSessionMetadata: {
          error: data.error,
          loading: data.loading
        },
        voteSession: data.voteSession,
        refetchVoteSession: data.refetch
      };
    }
  }),
  graphql(SectionsQuery, {
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          sectionsMetadata: {
            error: data.error,
            loading: data.loading
          }
        };
      }

      return {
        sectionsMetadata: {
          error: data.error,
          loading: data.loading
        },
        refetchSections: data.refetch,
        sections: data.sections
      };
    },
    skip: isNotInDiscussionAdmin
  }),
  graphql(LegalContentsQuery, {
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          legalContentsMetadata: {
            error: data.error,
            loading: data.loading
          }
        };
      }

      return {
        legalContentsMetadata: {
          error: data.error,
          loading: data.loading
        },
        refetchLegalContents: data.refetch,
        legalContents: data.legalContents
      };
    },
    skip: isNotInDiscussionAdmin
  }),
  graphql(TimelineQuery, {
    options: ({ locale }) => ({
      variables: { lang: locale }
    }),
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          timelineMetadata: {
            error: data.error,
            loading: data.loading
          }
        };
      }

      return {
        timelineMetadata: {
          error: data.error,
          loading: data.loading
        },
        refetchTimeline: data.refetch,
        timeline: data.timeline
      };
    },
    skip: props => isNotInLandingPageAdmin(props) && isNotInDiscussionAdmin(props)
  }),
  graphql(LandingPageModules, {
    options: ({ locale }) => ({
      variables: { lang: locale }
    }),
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          landingPageModulesMetadata: {
            error: data.error,
            loading: data.loading
          },
          landingPageModules: []
        };
      }

      return {
        landingPageModulesMetadata: {
          error: data.error,
          loading: data.loading
        },
        refetchLandingPageModules: data.refetch,
        landingPageModules: data.landingPageModules
      };
    },
    skip: isNotInLandingPageAdmin
  }),
  graphql(TextFields, {
    options: ({ locale }) => ({
      variables: { lang: locale }
    }),
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          textFieldsMetadata: {
            error: data.error,
            loading: data.loading
          }
        };
      }

      return {
        textFieldsMetadata: {
          error: data.error,
          loading: data.loading
        },
        refetchTextFields: data.refetch,
        textFields: data.textFields
      };
    },
    skip: props => isNotInDiscussionAdmin(props) || props.router.getCurrentLocation().search !== '?section=3'
  }),
  mergeLoadingAndError([
    'voteSessionMetadata',
    'sectionsMetadata',
    'legalContentsMetadata',
    'timelineMetadata',
    'landingPageModulesMetadata',
    'textFieldsMetadata'
  ]),
  manageErrorAndLoading({ displayLoader: true })
)(Administration);