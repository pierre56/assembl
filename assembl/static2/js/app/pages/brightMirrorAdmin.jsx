// @flow
import React from 'react';
import { I18n } from 'react-redux-i18n';

import SectionTitle from '../components/administration/sectionTitle';

const BrightMirrorAdmin = () => (
  <div className="bright-mirror-admin admin-box admin-content">
    <SectionTitle
      title={I18n.t('administration.brightMirrorSection.configureTitle')}
      annotation={I18n.t('administration.timelineAdmin.annotation')}
    />
  </div>
);

export default BrightMirrorAdmin;