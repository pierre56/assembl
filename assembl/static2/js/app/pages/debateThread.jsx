// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import Loader from '../components/common/loader';
import Ideas from '../components/debate/tableOfIdeas/ideas';
import AllIdeasQuery from '../graphql/AllIdeasQuery.graphql';

type Props = {
  phaseId: string,
  identifier: string,
  data: {
    loading: boolean,
    error: ?Error,
    rootIdea: Idea,
    ideas: Array<Idea>
  },
  params: {
    phaseId: string,
    themeId?: string,
    questionId?: string
  },
  children: React.Node
};

const DebateThread = ({ phaseId, identifier, data, params, children }: Props) => {
  if (!data) {
    return (
      <div className="debate">
        <Loader color="black" />
      </div>
    );
  }
  const { loading, ideas, rootIdea } = data;
  const themeId = params.themeId || null;
  const questionId = params.questionId || null;
  const isParentRoute = !(themeId || questionId) || false;
  const childrenElm = React.Children.map(children, child =>
    React.cloneElement(child, {
      id: themeId || questionId,
      identifier: identifier,
      phaseId: phaseId
    })
  );

  return (
    <div className="debate">
      {loading && isParentRoute && <Loader color="black" />}
      {!loading &&
        ideas &&
        isParentRoute && <Ideas key={phaseId} ideas={ideas} rootIdeaId={rootIdea.id} identifier={identifier} phaseId={phaseId} />}
      {!isParentRoute && <section className="debate-section">{childrenElm}</section>}
    </div>
  );
};

DebateThread.defaultProps = {
  data: null
};

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug
});

export default compose(
  connect(mapStateToProps),
  graphql(AllIdeasQuery, {
    skip: ({ discussionPhaseId }) => !discussionPhaseId
  })
)(DebateThread);