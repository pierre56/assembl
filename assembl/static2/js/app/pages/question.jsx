// @flow
import React from 'react';
import { connect } from 'react-redux';
import { compose, graphql } from 'react-apollo';
import { Translate } from 'react-redux-i18n';
import { Grid } from 'react-bootstrap';
import { Link } from 'react-router';

import manageErrorAndLoading from '../components/common/manageErrorAndLoading';
import Header from '../components/common/header';
import Posts from '../components/debate/survey/posts';
import Question from '../graphql/QuestionQuery.graphql';
import { get as getRoute } from '../utils/routeMap';
import HeaderStatistics, { statContributions, statMessages, statParticipants } from '../components/common/headerStatistics';
import { getIsPhaseCompletedById } from '../utils/timeline';

type NavigationParams = {
  questionIndex: string,
  questionId: string
};

type Props = {
  phaseId: string,
  timeline: Timeline,
  title: string,
  numContributors: number,
  numPosts: number,
  params: NavigationParams,
  thematicTitle: string,
  thematicId: string,
  imgUrl: string,
  slug: string,
  totalSentiments: number
};

export function DumbQuestion(props: Props) {
  const {
    phaseId,
    imgUrl,
    timeline,
    title,
    numContributors,
    numPosts,
    thematicTitle,
    thematicId,
    params,
    slug,
    totalSentiments
  } = props;
  const link = `${getRoute('idea', { slug: slug, phase: 'survey', phaseId: phaseId, themeId: thematicId })}`;
  let statElements = [];
  const numContributions = numPosts + totalSentiments;
  statElements = [statMessages(numPosts), statContributions(numContributions), statParticipants(numContributors)];
  const isPhaseCompleted = getIsPhaseCompletedById(timeline, phaseId);
  return (
    <div className="question">
      <div className="relative">
        <Header title={thematicTitle} imgUrl={imgUrl} phaseId={phaseId}>
          <HeaderStatistics statElements={statElements} />
        </Header>
        <div className="background-light proposals">
          <Grid fluid className="background-light">
            <div className="max-container">
              <div className="margin-xl">&nbsp;</div>
              <Link to={link} className="button-diamond-dark button-submit button-dark">
                <Translate value="debate.question.backToQuestions" />
                <span className="button-diamond-dark-back" />
              </Link>
              <div className="question-title">
                <div className="title-hyphen">&nbsp;</div>
                <h1 className="dark-title-1">
                  <Translate value="debate.survey.proposalsTitle" />
                </h1>
              </div>
              <div className="center">
                <h3 className="collapsed-title">
                  <span>{`${params.questionIndex}/ ${title}`}</span>
                </h3>
                <Posts questionId={params.questionId} themeId={thematicId} isPhaseCompleted={isPhaseCompleted} />
                <div className="back-btn-container">
                  <Link to={link} className="button-submit button-dark">
                    <Translate value="debate.question.backToQuestions" />
                  </Link>
                </div>
              </div>
              <div className="margin-xl">&nbsp;</div>
            </div>
          </Grid>
        </div>
      </div>
    </div>
  );
}

const mapStateToProps = state => ({
  lang: state.i18n.locale,
  slug: state.debate.debateData.slug,
  timeline: state.timeline
});

export default compose(
  connect(mapStateToProps),
  graphql(Question, {
    props: ({ data }) => {
      if (data.error || data.loading) {
        return {
          error: data.error,
          loading: data.loading
        };
      }

      const { question: { numContributors, numPosts, totalSentiments, parent: { img, title, id } } } = data;

      return {
        error: data.error,
        loading: data.loading,
        title: data.question.title,
        imgUrl: img ? img.externalUrl : '',
        numContributors: numContributors,
        numPosts: numPosts,
        thematicTitle: title,
        thematicId: id,
        totalSentiments: totalSentiments
      };
    }
  }),
  manageErrorAndLoading({ color: 'black', displayLoader: true })
)(DumbQuestion);