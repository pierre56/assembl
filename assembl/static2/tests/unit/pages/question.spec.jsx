import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbQuestion } from '../../../js/app/pages/question';

describe('Question page', () => {
  const props = {
    imgUrl: 'https://foo.bar/umgurl',
    isModerating: false,
    timeline: [
      {
        id: 'phaseFoo',
        title: 'Survey',
        identifier: 'survey',
        start: '2018-09-27T20:00:00+00:00',
        end: '2100-01-27T20:00:00+00:00'
      }
    ],
    title: 'Foo',
    thematicTitle: 'Bar',
    thematicId: 'ThematicId',
    numPosts: 2,
    numContributors: 2,
    totalSentiments: 1,
    params: {
      questionIndex: 'Index',
      questionId: 'FooInd'
    },
    slug: 'FooSlug',
    phaseId: 'phaseFoo'
  };

  it('should display a question', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<DumbQuestion {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });

  it('should display a question in moderation context', () => {
    const renderer = new ShallowRenderer();
    renderer.render(<DumbQuestion {...props} isModerating />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});