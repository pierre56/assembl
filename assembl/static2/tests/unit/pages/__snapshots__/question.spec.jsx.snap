// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question page should display a question 1`] = `
<div
  className="question"
>
  <div
    className="relative"
  >
    <Connect(Header)
      imgUrl="https://foo.bar/umgurl"
      phaseId="phaseFoo"
      title="Bar"
    >
      <HeaderStatistics
        statElements={
          Array [
            Object {
              "iconName": "message",
              "metricNameTranslateKey": "home.messages",
              "metricValue": 2,
            },
            Object {
              "iconName": "sentiment-neutral",
              "metricNameTranslateKey": "home.contribution",
              "metricValue": 3,
            },
            Object {
              "iconName": "profil",
              "metricNameTranslateKey": "home.participant",
              "metricValue": 2,
            },
          ]
        }
      />
    </Connect(Header)>
    <div
      className="background-light proposals"
    >
      <Grid
        bsClass="container"
        className="background-light"
        componentClass="div"
        fluid={true}
      >
        <div
          className="max-container"
        >
          <div
            className="margin-xl"
          >
             
          </div>
          <Link
            className="button-diamond-dark button-submit button-dark"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/FooSlug/debate/survey/theme/ThematicId"
          >
            <Translate
              value="debate.question.backToQuestions"
            />
            <span
              className="button-diamond-dark-back"
            />
          </Link>
          <div
            className="question-title"
          >
            <div
              className="title-hyphen"
            >
               
            </div>
            <h1
              className="dark-title-1"
            >
              <Translate
                value="debate.survey.proposalsTitle"
              />
            </h1>
          </div>
          <div
            className="center"
          >
            <h3
              className="collapsed-title"
            >
              <span>
                Index/ Foo
              </span>
            </h3>
            <Connect(withRouter(Apollo(Component)))
              isModerating={false}
              isPhaseCompleted={false}
              questionId="FooInd"
              themeId="ThematicId"
            />
            <div
              className="back-btn-container"
            >
              <Link
                className="button-submit button-dark"
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/FooSlug/debate/survey/theme/ThematicId"
              >
                <Translate
                  value="debate.question.backToQuestions"
                />
              </Link>
            </div>
          </div>
          <div
            className="margin-xl"
          >
             
          </div>
        </div>
      </Grid>
    </div>
  </div>
</div>
`;

exports[`Question page should display a question in moderation context 1`] = `
<div
  className="question"
>
  <div
    className="relative"
  >
    <Connect(Header)
      imgUrl="https://foo.bar/umgurl"
      phaseId="phaseFoo"
      title="Bar"
    >
      <HeaderStatistics
        statElements={
          Array [
            Object {
              "iconName": "message",
              "metricNameTranslateKey": "home.messages",
              "metricValue": 2,
            },
            Object {
              "iconName": "sentiment-neutral",
              "metricNameTranslateKey": "home.contribution",
              "metricValue": 3,
            },
            Object {
              "iconName": "profil",
              "metricNameTranslateKey": "home.participant",
              "metricValue": 2,
            },
          ]
        }
      />
    </Connect(Header)>
    <div
      className="background-light proposals"
    >
      <Grid
        bsClass="container"
        className="background-light"
        componentClass="div"
        fluid={true}
      >
        <div
          className="max-container"
        >
          <div
            className="margin-xl"
          >
             
          </div>
          <Link
            className="button-diamond-dark button-submit button-dark"
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/FooSlug/debate/survey/theme/ThematicId"
          >
            <Translate
              value="debate.question.backToQuestions"
            />
            <span
              className="button-diamond-dark-back"
            />
          </Link>
          <div
            className="question-title"
          >
            <div
              className="title-hyphen"
            >
               
            </div>
            <h1
              className="dark-title-1"
            >
              <Translate
                value="debate.survey.moderateProposalsTitle"
              />
            </h1>
          </div>
          <div
            className="center"
          >
            <h3
              className="collapsed-title"
            >
              <span>
                Index/ Foo
              </span>
            </h3>
            <Connect(withRouter(Apollo(Component)))
              isModerating={true}
              isPhaseCompleted={false}
              questionId="FooInd"
              themeId="ThematicId"
            />
            <div
              className="back-btn-container"
            >
              <Link
                className="button-submit button-dark"
                onlyActiveOnIndex={false}
                style={Object {}}
                to="/FooSlug/debate/survey/theme/ThematicId"
              >
                <Translate
                  value="debate.question.backToQuestions"
                />
              </Link>
            </div>
          </div>
          <div
            className="margin-xl"
          >
             
          </div>
        </div>
      </Grid>
    </div>
  </div>
</div>
`;
